Source/WebCore/platform/mediastream/gstreamer/RealtimeOutgoingMediaSourceGStreamer.cpp

```diff
@@ -609,37 +667,39 @@ void RealtimeOutgoingMediaSourceGStreamer::stopUpdatingStats()
 
 void RealtimeOutgoingMediaSourceGStreamer::teardown()
 {
+    GST_DEBUG_OBJECT(m_bin.get(), "Tearing down");
     if (m_transceiver)
         g_signal_handlers_disconnect_by_data(m_transceiver.get(), this);
 
-    stopOutgoingSource();
-    stopUpdatingStats();
+    stopOutgoingSource([&] {
+        stopUpdatingStats();
 
-    if (GST_IS_PAD(m_webrtcSinkPad.get())) {
-        auto srcPad = adoptGRef(gst_element_get_static_pad(m_bin.get(), "src"));
-        if (gst_pad_unlink(srcPad.get(), m_webrtcSinkPad.get())) {
-            GST_DEBUG_OBJECT(m_bin.get(), "Removing webrtcbin pad %" GST_PTR_FORMAT, m_webrtcSinkPad.get());
-            if (auto parent = adoptGRef(gst_pad_get_parent_element(m_webrtcSinkPad.get())))
-                gst_element_release_request_pad(parent.get(), m_webrtcSinkPad.get());
+        if (GST_IS_PAD(m_webrtcSinkPad.get())) {
+            auto srcPad = adoptGRef(gst_element_get_static_pad(m_bin.get(), "src"));
+            if (gst_pad_unlink(srcPad.get(), m_webrtcSinkPad.get())) {
+                GST_DEBUG_OBJECT(m_bin.get(), "Removing webrtcbin pad %" GST_PTR_FORMAT, m_webrtcSinkPad.get());
+                if (auto parent = adoptGRef(gst_pad_get_parent_element(m_webrtcSinkPad.get())))
+                    gst_element_release_request_pad(parent.get(), m_webrtcSinkPad.get());
+            }
         }
-    }
 
-    gst_element_set_locked_state(m_bin.get(), TRUE);
-    gst_element_set_state(m_bin.get(), GST_STATE_NULL);
-    if (auto pipeline = adoptGRef(gst_element_get_parent(m_bin.get())))
-        gst_bin_remove(GST_BIN_CAST(pipeline.get()), m_bin.get());
-    gst_element_set_locked_state(m_bin.get(), FALSE);
-
-    m_packetizers.clear();
-
-    m_bin.clear();
-    m_tee.clear();
-    m_rtpFunnel.clear();
-    m_allowedCaps.clear();
-    m_transceiver.clear();
-    m_sender.clear();
-    m_webrtcSinkPad.clear();
-    m_parameters.reset();
+        gst_element_set_locked_state(m_bin.get(), TRUE);
+        gst_element_set_state(m_bin.get(), GST_STATE_NULL);
+        if (auto pipeline = adoptGRef(gst_element_get_parent(m_bin.get())))
+            gst_bin_remove(GST_BIN_CAST(pipeline.get()), m_bin.get());
+        gst_element_set_locked_state(m_bin.get(), FALSE);
```

Comment:
I wonder if we should create an external locker for this.

Reply:
OK I can follow-up on that.

I'm not sure where you think the commit msg has a parse error. Grammar looks OK :)